components/magicui/typing-animation.tsx

```diff
@@ -0,0 +1,90 @@
+"use client";
+
+import { cn } from "@/lib/utils";
+import { motion, MotionProps } from "motion/react";
+import { useEffect, useRef, useState } from "react";
+
+interface TypingAnimationProps extends MotionProps {
+  children: string;
+  className?: string;
+  duration?: number;
+  delay?: number;
+  as?: React.ElementType;
+  startOnView?: boolean;
+}
+
+export function TypingAnimation({
+  children,
+  className,
+  duration = 100,
+  delay = 0,
+  as: Component = "div",
+  startOnView = false,
+  ...props
+}: TypingAnimationProps) {
+  const MotionComponent = motion.create(Component, {
+    forwardMotionProps: true,
+  });
+
+  const [displayedText, setDisplayedText] = useState<string>("");
+  const [started, setStarted] = useState(false);
+  const elementRef = useRef<HTMLElement | null>(null);
+
+  useEffect(() => {
+    if (!startOnView) {
+      const startTimeout = setTimeout(() => {
+        setStarted(true);
+      }, delay);
+      return () => clearTimeout(startTimeout);
+    }
+
+    const observer = new IntersectionObserver(
```

Comment:
What is this observer doing?

Reply:
in theory, it is making sure the text is in view (10% of viewport) before the animation is triggered to start. i guess that's not necessary here - cause anytime the page is refreshed the hero is going to be in the viewport like 100%... might be worth keeping this component unchanged if we want to use the animated text in some other parts of the site that aren't automatically in view ? it's also the default from magic ui, so keeping it the same will be simpler for other implementations...